### ISSUE TITLE
Build failed on MacBook Pro M1

### ISSUE THREAD

**Issue by harisalam** [opensearch-project/k-NN#580]
**Updated:** 
Please follow this to build k-NN in M1 for now to unblock yourself. Better process should be implemented still.

OpenSearch 3.0.0 distribution for mac does not exist in the public url. You can lower the version of k-NN to 2.3.0 which will use OpenSearch 2.3.0 release. Or, you can build your own OpenSearch distribution and use it as follow.

1. Fork the [OpenSearch Repo](https://github.com/opensearch-project/OpenSearch) into your github account.
2. Checkout the OpenSearch Repo in your laptop using right github commands.
3. Run the below commands, as provided in the [repo](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#build).
$> cd OpenSearch && ./gradlew assemble && ./gradlew localDistro
4. You should see a opensearch-min-3.0.0-SNAPSHOT-darwin-x64.tar.gz file present in distribution/archives/darwin-tar/build/distributions/
5. Build k-NN by passing the OpenSearch distribution path in `$> ./gradlew <integTest/run> -PcustomDistributionUrl="<Full path to .tar.gz file you noted above>"`
---------------------

With default setting of Make file I was getting the following error:
    clang: error: the clang compiler does not support '-march=native'
Then someone from the community also experience an issue with this and solution was provided by changing -march=native to -mcpu=apple-a14, but that landed me on another issue which is omp.h not found. After researching for quite a long time I found out this error is caused of clang incompatibility on apple-m1.

> Task :buildJniLib FAILED
Consolidate compiler generated dependencies of target NonMetricSpaceLib
[ 88%] Built target NonMetricSpaceLib
Consolidate compiler generated dependencies of target opensearchknn_common
[ 94%] Built target opensearchknn_common
Consolidate compiler generated dependencies of target opensearchknn_nmslib
[100%] Built target opensearchknn_nmslib
[  2%] Building CXX object external/faiss/faiss/CMakeFiles/faiss.dir/AutoTune.cpp.o
In file included from /Users/harisalam/Documents/k-NN/jni/external/faiss/faiss/AutoTune.cpp:34:
In file included from /Users/harisalam/Documents/k-NN/jni/external/faiss/faiss/IndexHNSW.h:14:
/Users/harisalam/Documents/k-NN/jni/external/faiss/faiss/impl/HNSW.h:16:10: fatal error: 'omp.h' file not found
#include <omp.h>
         ^~~~~~~
1 error generated.
make[3]: *** [external/faiss/faiss/CMakeFiles/faiss.dir/AutoTune.cpp.o] Error 1
make[2]: *** [external/faiss/faiss/CMakeFiles/faiss.dir/all] Error 2
make[1]: *** [CMakeFiles/opensearchknn_faiss.dir/rule] Error 2
make: *** [opensearchknn_faiss] Error 2

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':buildJniLib'.
> Process 'command 'make'' finished with non-zero exit value 2

**Comment by nknize** [comment]
I don't have a M1 or even a Mac to try this out so I'm largely probing here. You may have already tried this, but did you try the `/opt/homebrew` bundled clang which looks to [support the M1 as of 3.0.0](https://brew.sh/2021/02/05/homebrew-3.0.0/)? The M1 discussion is in a few different repos ([here's a good entry point](https://github.com/opensearch-project/opensearch-build/issues/37)) so maybe someone there can also help move this forward? @bbarani @dblock? Also  /cc @jmazanec15 @martin-gaievski, has anyone on K-nn successfully built the native faiss library on the M1? Looks like apple's clang gives issues but bundled homebrew clang works?

**Comment by martin-gaievski** [comment]
I think @heemin was facing issues with build on m1

**Comment by heemin32** [comment]
> I think @heemin32 was facing issues with build on m1

I haven't tried to build k-NN. I was having issue with running bwc integTest of OpenSearch.

**Comment by jmazanec15** [comment]
@harisalam I dont have M1 mac, but Ill see if I can get access to one.

Have you tried:
```
brew install gcc
```

you might be able to get passed omp failure this way, because gcc toolchain will ship an implementation of omp. This works for me on mac x86 - not sure if itll work on M1.

**Comment by heemin32** [comment]
After changing -march=native to -mcpu=apple-m1, I see following error.
```
/Users/heemin/workplace/OS/k-NN/jni/external/nmslib/similarity_search/src/distcomp_scalar.cc:85:9: error: pragma message requires parenthesized string
#pragma message WARN("ScalarProductSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
```

**Comment by harisalam** [comment]
> @harisalam I dont have M1 mac, but Ill see if I can get access to one.
> 
> Have you tried:
> 
> ```
> brew install gcc
> ```
> 
> you might be able to get passed omp failure this way, because gcc toolchain will ship an implementation of omp. This works for me on mac x86 - not sure if itll work on M1.

@jmazanec15 I have tried that but still its wasn't able to find omp.h library, then I figured that out by manually passing the location of omp.h library in libomp directory. After that I landed on another error which was related to clang "ld: symbol(s) not found for architecture arm64".

**Comment by harisalam** [comment]
> After changing -march=native to -mcpu=apple-m1, I see following error.
> 
> ```
> /Users/heemin/workplace/OS/k-NN/jni/external/nmslib/similarity_search/src/distcomp_scalar.cc:85:9: error: pragma message requires parenthesized string
> #pragma message WARN("ScalarProductSIMD<float>: SSE2 is not available, defaulting to pure C++ implementation!")
> ```

@heemin32 the solution for that scenario was to comment out pragma errors in CMakeList file. But still it could build the project successfully

**Comment by harisalam** [comment]
> I don't have a M1 or even a Mac to try this out so I'm largely probing here. You may have already tried this, but did you try the `/opt/homebrew` bundled clang which looks to [support the M1 as of 3.0.0](https://brew.sh/2021/02/05/homebrew-3.0.0/)? The M1 discussion is in a few different repos ([here's a good entry point](https://github.com/opensearch-project/opensearch-build/issues/37)) so maybe someone there can also help move this forward? @bbarani @dblock? Also /cc @jmazanec15 @martin-gaievski, has anyone on K-nn successfully built the native faiss library on the M1? Looks like apple's clang gives issues but bundled homebrew clang might work?

@nknize I have also tried that by following the developer guide, but it wasn't successfully building for the following reasons:
1.  march=native error
2. omp.h not found
3. clang error "ld: symbol(s) not found for architecture arm64"

**Comment by harisalam** [comment]
> 

1. March native error was addressed by making changes to the CMake file by simply changing -march=native to -mcpu=apple-a14
2. omp.h was addressed by specifying the location to omp.h file on the local drive
3. Have no idea how to deal with clang error; this error is most probably for one specific reason which is the system tries to execute .cpp file using clang however clang++ needs to used for execution or compilation.

**Comment by heemin32** [comment]
I got stuck with the clang error `ld: symbol(s) not found for architecture arm64` as well but @jmazanec15 helped me out here. Thanks @jmazanec15

```
// Go to jni folder
cd k-NN/jni

// File changes required
sed -i -e 's/\/usr\/local\/opt\/libomp\//\/opt\/homebrew\/opt\/llvm\//g' CMakeLists.txt
sed -i -e 's/-march=native/-mcpu=apple-m1/g' external/nmslib/similarity_search/CMakeLists.txt
sed -i -e 's/pragma message WARN/pragma message /g' external/nmslib/similarity_search/src/distcomp_scalar.cc
sed -i -e 's/-mcpu=apple-a14/-mcpu=apple-m1/g' external/nmslib/python_bindings/setup.py
sed -i -e 's/__aarch64__/__undefine_aarch64__/g' external/faiss/faiss/utils/distances_simd.cpp

// Install llvm
brew install llvm
echo 'export PATH="/opt/homebrew/opt/llvm/bin:$PATH"' >> ~/.zshrc
source ~/.zshrc

// Set compiler path for CMAKE
export CC=/opt/homebrew/opt/llvm/bin/clang
export CXX=/opt/homebrew/opt/llvm/bin/clang++

// Build
cmake . --fresh
make
```

**Comment by harisalam** [comment]
@heemin32 With above changes to the files I have been able to override clang errors, however, build is continuously failing because the mentioned snapshots aren't accessible for some reason.

<img width="1512" alt="Screen Shot 2022-10-30 at 5 03 49 PM" src="https://user-images.githubusercontent.com/79714825/198903818-c994ea1a-4fa0-4c87-81be-02279242bef1.png">

**Comment by harisalam** [comment]
@heemin32 @jmazanec15 When I try to access (https://artifacts.opensearch.org/snapshots/core/opensearch/3.0.0-SNAPSHOT/opensearch-min-3.0.0-SNAPSHOT-darwin-arm64-latest.tar.gz) using https it throws me an error "Access Denied". Are there any bugs in the new release which is causing this issue?
<img width="1506" alt="Screen Shot 2022-10-30 at 6 30 48 PM" src="https://user-images.githubusercontent.com/79714825/198907310-9e29cf9e-ee73-4b33-9f59-6dbac5e7fae5.png">

**Comment by heemin32** [comment]
OpenSearch 3.0.0 distribution for mac does not exist in the public url. You can lower the version of k-NN to 2.3.0 which will use OpenSearch 2.3.0 release. Or, you can build your own OpenSearch distribution and use it as follow.

1. Fork the [OpenSearch Repo](https://github.com/opensearch-project/OpenSearch) into your github account.
2. Checkout the OpenSearch Repo in your laptop using right github commands.
3. Run the below commands, as provided in the [repo](https://github.com/opensearch-project/OpenSearch/blob/main/DEVELOPER_GUIDE.md#build).
$> cd OpenSearch && ./gradlew assemble && ./gradlew localDistro
4. You should see a opensearch-min-3.0.0-SNAPSHOT-darwin-x64.tar.gz file present in distribution/archives/darwin-tar/build/distributions/
5. Build k-NN by passing the OpenSearch distribution path in `$> ./gradlew <integTest/run> -PcustomDistributionUrl="<Full path to .tar.gz file you noted above>"`

**Comment by navneet1v** [comment]
@heemin32 can you check if darwin arm64 snapshot is present or not?

**Comment by austintlee** [comment]
The workaround mentioned above works for me on the 2023 M2 Pro and I also have the issue with the missing darwin-arm64 on 'integTest'.

**Comment by austintlee** [comment]
I can build k-nn, but I can't install in in my local repo.  I get this error:
```
Execution failed for task ':publishNebulaPublicationToMavenLocal'.
> Failed to publish publication 'nebula' to repository 'mavenLocal'
   > Invalid publication 'nebula': supplied groupId (org.opensearch) does not match value from POM file (org.opensearch.plugin). Cannot edit groupId directly in the POM file.
```

**Comment by TrungBui59** [comment]
Hi everyone, how can you guys fix this issue? I followed the instructions by @heemin32  but found the issue of cannot find omp.h:

```
/Users/trungbui/k-NN/jni/external/faiss/faiss/impl/HNSW.h:16:10: fatal error: 'omp.h' file not found
#include <omp.h>
```

**Comment by navneet1v** [comment]
@ryanbogan can you help here.. you recently work on k-nn with M1 chip

**Comment by bbarani** [comment]
You can download the Darwin snapshot -min build of OpenSearch [here](https://artifacts.opensearch.org/snapshots/core/opensearch/3.0.0-SNAPSHOT/opensearch-min-3.0.0-SNAPSHOT-darwin-x64-latest.tar.gz ) @heemin32 @navneet1v

**Comment by heemin32** [comment]
Closing it in favor of https://github.com/opensearch-project/k-NN/pull/1302
